gslib/commands/iam.py

```diff
@@ -480,7 +480,7 @@ def _PatchIam(self):
     # expecting to come across the -r, -f flags here.
     it = iter(self.args)
     for token in it:
-      if STORAGE_URI_REGEX.match(token):
+      if IsBucketNameValid(token):
```

Comment:
I thought about this a bit afterward, and I think we should be doing minimal validation on the client side - if it's an invalid bucket name, URI, etc., then we should pass it along to the service via the appropriate API call and let the returned error message explain that the given resource name was invalid or didn't exist.

Thus, to check if something is a valid storage URI or not, we should just check for the presence of some `[a-z]` character(s), followed by `://`, followed by at least one more character. So basically what we had before, except change the regex to be `r'[a-z]+://.+'`.

wdyt?

Reply:
That sounds pretty reasonable to me since we only really need to distinguish the bucket strings from binding strings. Assuming bindings are always `:` delimited with two to three tokens, and bucket strings are always two tokens delimited by `://` is there any case where it would be valid for any part of the binding to start with `//`? Fearful for a case of falsely identifying a binding as a bucket. 

Particularly this might be a problem for the domain tag, where the example lists `domain:www.my-domain.org:objectViewer` someone might do `domain:https://my-domain.org:objectViewer` which would ID as a bucket string. I'm not sure if the latter is supposed to be valid binding syntax, but it seems like a reasonably possible thing for sometime to do and get a weird error from.

We know that neither the bucket prefix or URI will contain colons, so we could use your regex, plus a rule "contains no more than one colon" possibly. I'm testing and struggling to make a regex that satisfies that without producing partial matches.

We could combine them with something like:
```
r = re.compile(r'[a-z]+://.+')

if r.match(token) and token.count(':') < 2:
   do bucket stuff
```

or the regex is getting simple enough that we might be able to do without a regex at all:

```
if '://' in token and token.count(':') < 2:
    do bucket stuff
```

Do either of those two snippets sound reasonable?

Reply:
(made some edits to above comment)

Reply:
I think you're thinking of `re.search()`; the above command uses `re.match()`, which only returns a match if there's an exact match starting from the beginning of the string; thus the beginning of the string would have to be the `[a-z]` chars, followed by `://`.  Something with a colon before the `://`, like `domain:foo.bar:someRole` wouldn't match. :)

Reply:
Oh oh! You're right! That is much simpler then. I'll add your regex and cut the function :)

Reply:
Ok! Put the new regex in, and pulled the validation function out. 👍